### ISSUE TITLE
RMSE  calculation

### ISSUE THREAD

**Issue by RomanSteinberg** [catboost/catboost#150]
This is minimal [example](https://gist.github.com/RomanSteinberg/956d57b5ee22af1a47da23f5e9e4d1f3) I wish to discuss. I used Python 2.7, catboost 0.2.5 to obtain results in the gist.

1. According to [score](https://tech.yandex.com/catboost/doc/dg/concepts/python-reference_catboostregressor_score-docpage/) routine documentation, it calculates MSE. I checked it and according to `root score 167.908479` calculated as a root from catboost score and `rmse 167.908479` calculated using numpy, it is really MSE. Why documentation links to loss functions? Why CatBoostRegressor constructed with RMSE loss function (or any other) provides score interface for MSE only?
2. (resolved) According to my example, CatBoostRegressor calculates very strange scores (errors) during training. Compare `bestTest = 40.44999897` and `mse 28193.257265, rmse 167.908479`. So, the measured metric is not one of MSE, RMSE. What is it? 
3. According to [get_test_eval](https://tech.yandex.com/catboost/doc/dg/concepts/python-reference_catboostregressor_get_test_eval-docpage/) routine documentation, it returns the values of the evaluation loss functions. If you look at my example, it returns predictions in 2d-array. Look at `Predictions:  [-162.90847884]  get_test_eval:  [[-162.90847883612867]]`. Documentation or implementation should be fixed, I think.
4. According to [predict](https://tech.yandex.com/catboost/doc/dg/concepts/python-reference_catboostregressor_predict-docpage/) routine documentation, it provides weight parameter. I use weights to rebalance my dataset. For example, for objects of rare classes in classification task. When we use predict for new objects we don't have any information about their weights, because we know nothing about classes. If you look at [scikit-learn interfaces](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html), they have no weight parameter for predict. So, why you need it?  
5. RMSE calculation. When I provide weights for training I expect to get training error which calculated with respect to them. But test error should not. As for me weights is a good way not to clone some objects in dataset. May be there are  some other cases which I don't know. Then share those cases, please.

I created this example because I had very strange numbers when I trained model with weigths using catboost GPU version. It was much easier to create this example than for GPU version, but I tried to analyze cpp code for RMSE calculation and it seems wrong to me. Please, check it.

As for now, I stoped to use weights at all and clone objects in my dataset.

UPD: Problem 2 according to Noxoomo comment.

**Comment by RomanSteinberg** [comment]
@Noxoomo consider my example, please. I calculate weighted RMSE using this formula, catboost not.

**Comment by Noxoomo** [comment]
In your example wrmse = np.sqrt(np.sum(weights * (pred - test_labels) ** 2)/np.linalg.norm(weights))

np.linalg.norm(weights) gives l2-norm of weights. So your normalization is different.

**Comment by RomanSteinberg** [comment]
Sorry, I made a mistake. I fixed it in gist. And now I still have no idea how catboost calculates weighted RMSE, because it doesn't fit your formula.

**Comment by Noxoomo** [comment]
Now everything looks good. CatBoost prints score from last iteration and they are equal to numpy-calculated. (Train: mse 14355.246379, rmse 119.813381, wrmse 49.965634  equal to 9:	learn 49.9656332; test 167.9084788 equal to Test: mse 28193.257265, rmse 167.908479)

To get bestTest score in your example you should add use_best_model=True for CatBoostRegressor. CatBoost don't make model shrinkage by default (and uses all 10 iterations, in your example best iteration was first).

**Comment by RomanSteinberg** [comment]
@Noxoomo problem 2 resolved, but not others.

**Comment by RomanSteinberg** [comment]
Can you provide examples for both paragraphs in the previous post?

**Comment by annaveronika** [comment]
About 4: you are right, we don't use weights in predict, the argument is just ignored. We'll remove it soon, thanks for pointing this out.

**Comment by annaveronika** [comment]
About 5: there is a possibility to use weighted error on train and on test. If you don't want weighted error on test, you should specify weight == 1 for these objects. In other cases it is useful to specify weights for test.

**Comment by annaveronika** [comment]
About 1: thanks for pointing out, will fix it soon.

**Comment by annaveronika** [comment]
About 3: also coming soon, thanks!

**Comment by annaveronika** [comment]
Everything except for 5 is fixed, 5 we will not change.
Thanks a lot for your feedback!
